File: Practica 4/CivitasJava4/CivitasJava/src/civitasjava/SorpresaSalirCarcel.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitasjava;

import java.util.ArrayList;

/**
 *
 * @author francisco
 */
public class SorpresaSalirCarcel extends Sorpresa {
    
    private MazoSorpresas mazo;

    SorpresaSalirCarcel(MazoSorpresas mazo){  
        super("Evita la carcel");

        this.mazo = mazo;
    }
    
    /**
     * @brief Si nadie tiene la sorpresa para evitar la carcel, la obtine el jugador actual
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    @Override
    void aplicarAJugador (int actual, ArrayList<Jugador> todos){
      if( jugadorCorrecto(actual, todos) ){
        this.informe(actual, todos);
        
        // Preguntamos si alguien tiene la sorpresa para evitar la carcel
        boolean tieneSalvoconducto = false;
        for(int i = 0; i < todos.size() && !tieneSalvoconducto; i++)
            tieneSalvoconducto = todos.get(i).tieneSalvoconducto();
        
        // Si nadie la tiene, la obtiene el jugador actual
        if(!tieneSalvoconducto){
            todos.get(actual).obtenerSalvoconducto(this);
            this.salirDelMazo();
        }
        
        // Por que no miramos simplemente si alguien tiene la sorpresa 
        // haciendo mazo.cartasEspeciales.size() > 0 ??
      }
    }
    
    
    /**
     * @brief Inhabilita la carta especial en el mazo de sorpresa
     * @return void
     */
    void salirDelMazo(){
        this.mazo.inhabilitarCartaEspecial(this);
    }
    
    
    /**
     * @brief Habilita la carta especial en el mazo de sorpresas
     * @return void
     */
    void usada(){
        this.mazo.habilitarCartaEspecial(this);
    }
}

File: Practica 4/CivitasRuby4/CivitasRuby/lib/sorpresa_salir_carcel.rb
module Civitas

class Sorpresa_salir_carcel < Sorpresa
  
# @brief Constructor de la clase
# @param mazo, el mazo con en el que se juega
def initialize(mazo)
  super("Evita la carcel")

  @mazo = mazo
end


# @brief Si nadie tiene la sorpresa para evitar la carcel, la obtiene el
#   jugador actual
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void  
def aplicar_a_jugador(actual,todos)      
  if (jugador_correcto(actual,todos))
    informe(actual,todos)
    Diario.instance.ocurre_evento("Se procede a pedir el salvoconducto")
  end

  for i in 0..todos.size-1 do
    if (todos.at(i).tiene_salvoconducto)
      salvoconducto = true
    end
  end

  if (!salvoconducto)
    todos.at(actual).obtener_salvoconducto(self)
    salir_del_mazo
  end

  return salvoconducto
end


# @brief Inhabilita la carta especial en el mazo de sorpresas
# @return Es void
def salir_del_mazo()
   @mazo.inhabilitar_carta_especial(self)
end

# @brief Habilita la carta especial en el mazo de sorpresas
# @return Es void
def usada()
  @mazo.habilitar_carta_especial(self)
end
  

end

end